internal/controller/controllers/agent_controller.go

```diff
@@ -80,6 +81,9 @@ func (r *AgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 
 	// Retrieve clusterDeployment
 	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return r.deleteAgent(ctx, req.NamespacedName)
+		}
```

Comment:
Typically if you have cleanup to do the best practice is to make sure there's a finalizer on clusterdeployment in this case (that will change shortly, see below), and when you see ClusterDeployment with deletionTimestamp set and that finalizer still present, you know you need to take action.

This might be fine but there are some weird cases I can think of. Hive supports cluster migration from one Hive shard to another, in which case if the ClusterDeployment were gone, Agents would start deleting.
Or, it may also proc if you are creating Agents with a ref to a ClusterDeployment that doesn't exist yet, (maybe it's further down in file of yaml docs or a kubectl apply command). 

Just wanted to suggest the finalizer approach as that should be a more reliable way to handle this. This might be fine as is in which case disregard.

Once the AgentClusterInstall PR merges the place for such a finalizer would be on that resource, not ClusterDeployment.

Reply:
Totally agree. I was just waiting for Nir to elaborate on how this code is expected to work as-is before bringing up finalizers, but I agree this almost certainly calls for using one.

Reply:
Can you please point me to a good reference re: finalizers? in case we decide to go with that (just replied on the other comment).

Reply:
https://book.kubebuilder.io/reference/using-finalizers.html

Reply:
resolving this comment, since this is no longer a part of this PR.
to be included in a follow-up PR that handles host deregister.